internal/artifacts/module_arch.go

```diff
@@ -40,29 +41,242 @@ func ExecuteBuild(source, target string, extensions []string, moduleName, platfo
 }
 
 // ExecuteSoloBuild - executes build of module from stand alone command
-func ExecuteSoloBuild(source, target string, extensions []string, moduleName string, wdGetter func() (string, error)) error {
-	logs.Logger.Infof(buildMsg, moduleName)
+func ExecuteSoloBuild(source, target string, extensions []string, modulesNames []string, allDependencies bool, wdGetter func() (string, error)) error {
+
+	if len(modulesNames) == 0 {
+		return errors.New(buildFailedOnEmptyModulesMsg)
+	}
 
 	sourceDir, err := getSoloModuleBuildAbsSource(source, wdGetter)
 	if err != nil {
-		return errors.Wrapf(err, buildFailedMsg, moduleName)
+		return errors.Wrapf(err, getBuildMsg(buildFailedMsg, multiBuildFailedMsg, modulesNames))
 	}
 
-	targetDir, err := getSoloModuleBuildAbsTarget(sourceDir, target, moduleName, wdGetter)
+	loc, err := dir.Location(sourceDir, "", dir.Dev, extensions, wdGetter)
 	if err != nil {
-		return errors.Wrapf(err, buildFailedMsg, moduleName)
+		return errors.Wrapf(err, getBuildMsg(buildFailedMsg, multiBuildFailedMsg, modulesNames))
 	}
 
-	loc, err := dir.Location(sourceDir, targetDir, dir.Dev, extensions, wdGetter)
+	mtaObj, err := loc.ParseFile()
 	if err != nil {
-		return errors.Wrapf(err, buildFailedMsg, moduleName)
+		return err
+	}
+
+	if len(modulesNames) > 1 {
+		logs.Logger.Infof(multiBuildMsg, convertArrayToString(modulesNames))
+	}
+
+	err = checkBuildResultsConflicts(mtaObj, sourceDir, target, extensions, modulesNames, wdGetter)
+	if err != nil {
+		return errors.Wrapf(err, getBuildMsg(buildFailedMsg, multiBuildFailedMsg, modulesNames))
+	}
+
+	sortedModules, err := sortSelectedModules(mtaObj, modulesNames)
+	if err != nil {
+		return errors.Wrapf(err, getBuildMsg(buildFailedMsg, multiBuildFailedMsg, modulesNames))
+	}
+
+	err = buildSelectedModules(sourceDir, target, extensions, mtaObj, sortedModules, allDependencies, wdGetter)
```

Comment:
This is similar to what we discussed yesterday but I think it's worth doing.
Instead of sorting the modules by dependencies, and looping over each module and building its dependencies and then the module, it would be more efficient and I think also easier to understand if you get the list of modules with dependencies, order them (after you have all the dependencies) with `sortSelectedModules`, then build them by order.
It will also make it easier to implement the next step of limiting the dependent modules list, plus make sure you only build each module once.